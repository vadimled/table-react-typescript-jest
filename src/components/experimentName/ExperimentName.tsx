import React, { FC, useState } from "react";
import { ReactComponent as EditIcon } from "../../assets/edit.svg";
import { ReactComponent as ClearUndoIcon } from "../../assets/close.svg";
import { ReactComponent as SaveIcon } from "../../assets/save.svg";
import * as styles from "./ExperimentName.module.scss";
import { Input } from "antd";
import CustomIcon from "../customIcon";
import {
  ActionsType,
  RecordMainType,
  IActionWithoutPayload,
} from "../../types";

type PropsType = {
  id: string;
  name: string;
  isSaveNameLoading: boolean;
  insertedText: RecordMainType | null;
  onChangeEditName: (
    event: React.ChangeEvent<HTMLInputElement>
  ) => void | undefined;
  onClearUndo: () => IActionWithoutPayload | void;
  onClickSave: (data: RecordMainType | null) => ActionsType | void;
  onStartEditName: (data: RecordMainType) => ActionsType | void;
};

const ExperimentName: FC<PropsType> = ({
  id,
  name,
  isSaveNameLoading,
  onChangeEditName,
  onClearUndo,
  onClickSave,
  onStartEditName,
  insertedText,
}) => {
  const [isEdit, setMode] = useState(insertedText?.id === id);

  const onClickEditIcon = () => {
    setMode(true);
    onStartEditName({ id, name });
  };
  const onClickCleanIcon = () => {
    setMode(false);
    onClearUndo();
  };
  const onClickSaveIcon = () => {
    if (insertedText?.name !== name) {
      setMode(false);
      onClickSave(insertedText);
    }
  };

  const renderNameComponent = (): JSX.Element => {
    if (isEdit && insertedText) {
      return (
        <>
          <Input
            data-testid="input-name-component"
            placeholder="Edit name"
            onChange={onChangeEditName}
            value={insertedText?.name || ""}
            suffix={
              <ClearUndoIcon
                className="delete-icon"
                onClick={onClickCleanIcon}
              />
            }
          />
          <CustomIcon Icon={SaveIcon} onClick={onClickSaveIcon} />
        </>
      );
    } else {
      return (
        <>
          <div data-testid="edit-name-component" className="show-name-field">
            {name}
          </div>
          <CustomIcon
            Icon={EditIcon}
            loading={isSaveNameLoading}
            onClick={onClickEditIcon}
            disabled={!!insertedText && insertedText?.id !== id}
          />
        </>
      );
    }
  };

  return (
    <div className={styles["experiment-name-wrapper"]}>
      {renderNameComponent()}
    </div>
  );
};

export default ExperimentName;
