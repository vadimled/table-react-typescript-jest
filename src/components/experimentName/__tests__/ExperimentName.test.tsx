import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import ExperimentName from "../ExperimentName";

describe("ExperimentName", (): void => {
  const defaultProps = {
    id: "1",
    name: "name",
    isSaveNameLoading: false,
    onChangeEditName: jest.fn(),
    onClearUndo: jest.fn(),
    onClickSave: jest.fn(),
    onStartEditName: jest.fn(),
  };

  afterEach(cleanup);
  it("matches default ExperimentName snapshot", (): void => {
    const { asFragment } = render(
      <ExperimentName
        insertedText={{ id: "1", name: "name" }}
        {...defaultProps}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it("should render EditIcon", (): void => {
    const { getByTestId } = render(
      <ExperimentName insertedText={null} {...defaultProps} />
    );
    const editIcon = getByTestId("edit-name-component");
    expect(editIcon).toBeInTheDocument();
  });
  it("should render EditIcon with grey mode", (): void => {
    const { getByTestId } = render(
      <ExperimentName
        insertedText={{ id: "2", name: "name" }}
        {...defaultProps}
      />
    );
    const editIcon = getByTestId("target-icon");
    expect(editIcon).toHaveClass("current-icon disabled");
  });
  it("should render SaveIcon and input", (): void => {
    const { getByTestId } = render(
      <ExperimentName
        insertedText={{ id: "1", name: "name" }}
        {...defaultProps}
      />
    );

    const input = getByTestId("input-name-component");
    expect(input).toBeInTheDocument();
  });
  it("after click on SaveIcon shouldn't render input", (): void => {
    const { getByTestId } = render(
      <ExperimentName
        insertedText={{ id: "1", name: "name" }}
        {...defaultProps}
      />
    );
    const input = getByTestId("input-name-component");
    const customIconWrapper = getByTestId("custom-icon");
    fireEvent.click(customIconWrapper);
    expect(input).not.toBeInTheDocument();
  });
});
