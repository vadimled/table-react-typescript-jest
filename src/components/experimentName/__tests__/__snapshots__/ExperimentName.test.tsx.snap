// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExperimentName matches default ExperimentName snapshot 1`] = `
<DocumentFragment>
  <div>
    <span
      class="ant-input-affix-wrapper"
    >
      <input
        class="ant-input"
        data-testid="input-name-component"
        placeholder="Edit name"
        type="text"
        value="name"
      />
      <span
        class="ant-input-suffix"
      >
        <reactcomponent
          class="delete-icon"
        />
      </span>
    </span>
    <div
      data-testid="custom-icon"
    >
      <div
        class="current-icon"
        data-testid="target-icon"
      >
        <reactcomponent />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
